### PR TITLE
Add `ClientRequestContext` to the filters of RetryRule and CircuitBre…

### PR DESCRIPTION
**Pull request by ikhoon**
…akerRule

Motivation:
It is useful to use `ClientRequestContext` information to build a complex `RetryRule{WithContent}` and `CircuitBreakerRule{WithContent}`.
For example, if you apply a `RetryRule` only to a specific RPC method,
then you need to check the current served method using `ClientRequestContext#rpcRequest()`
Although it could be possible by implementing `RetryRule`, `RetryRuleBuilder` would be more convenient.
```java
Backoff backoff = ...
RetryRuleWithContent<RpcResponse> rule = (ctx, response, cause) -> {
    
    if (ctx.log().isAvailable(RequestLogProperty.REQUEST_HEADERS)) {
        ...
    }

    final RpcRequest request = ctx.rpcRequest();
    if (request != null && !safeMethods.contains(request.method())) {
        if (isRetryableException(cause)) {
            return CompletableFuture.completedFuture(RetryDecision.retry(backoff));
        }
    }
    return CompletableFuture.completedFuture(RetryDecision.next());
};

// Versus:
RetryRuleWithContent.<RpcResponse>builder()
                    .onServerErrorStatus()
                    .onException((ctx, ex) -> {
                        RpcRequest request = ctx.rpcRequest();
                        if (request != null && !safeMethods.contains(request.method())) {
                            return isRetryableException(throwable);
                        } else {
                            return false;
                        }
                    })
                    .thenBackoff(backoff);
```

Modifications:

- Add new methods that take `BiPredicate<ClientRequestContext, ?>` as a predicate.
- Deprecate on*(Predicate<?>) methods in the followings in favor of `on*(BiPredicate<ClientRequestContext, ?>)`:
  - RetryRule and RetryRuleBuilder
  - RetryRuleWithContent and RetryRuleWithContentBuilder
  - CircuitBreakerRule and CircuitBreakerRuleBuilder
  - CircuitBreakerRuleWithContent and CircuitBreakerRuleWithContentBuilder

Result:

You can now use ClientRequestContext to build a complex RetyRule and CircuitBreakerRule.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java [hunk 1]
```diff
@@ -46,55 +48,90 @@
...
+
+    /**
+     * Creates a new instance with the specified {@code requestHeadersFilter}.
+     *
+     * @deprecated Use {@link #AbstractRuleBuilder(BiPredicate)}.
+     */
+    @Deprecated
     protected AbstractRuleBuilder(Predicate<? super RequestHeaders> requestHeadersFilter) {
-        this.requestHeadersFilter = (Predicate<RequestHeaders>) requestHeadersFilter;
+        this(toBiPredicateForSecond(requireNonNull(requestHeadersFilter, "requestHeadersFilter")));
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:76]: How about `BiPredicates.second()`?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:76]: I thought that names too. I was not sure that the `second()` expresses the behavior. 🤔

- [Functions.second()](https://github.com/line/armeria/blob/ead4d1018db67aa68fe15e4ac237682f82662a61/core/src/main/java/com/linecorp/armeria/common/util/Functions.java#L207) returns a BiFunction that returns the second arguments.
- [BiPredicates.second(Predicate)] that returns a BiPredicate that uses only the second argument to evaluate the predicate.

How about `BiPredicates.ofSecond(Predicate)` at least to indicate the method is a factory?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:76]: We're probably gonna remove it once the deprecated methods are gone, so.. maybe OK not to care about naming? 😅

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:76]: > We're probably gonna remove it once the deprecated methods are gone, so.. maybe OK not to care about naming? 😅

Yes, the `BiPredicates` class will be removed eventually.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleBuilder.java:76]: If so, then it's fine as it is. 😄

**Pull request merged**